File: csharp/HiringSurvey.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Newtonsoft.Json;

namespace HiringSurvey
{
    class Program
    {
        const string DATA_FILE = "candidates.json";

        static void Main(string[] args)
        {
            while (true)
            {
                Console.WriteLine("\nMenu:");
                Console.WriteLine("1. Fill out the survey");
                Console.WriteLine("2. Search responses");
                Console.WriteLine("3. Exit");
                Console.Write("Choose an option: ");
                var choice = Console.ReadLine();

                if (choice == "1")
                {
                    var candidate = CollectCandidateInfo();
                    SaveCandidateInfo(candidate);
                    Console.WriteLine("Candidate information saved successfully.");
                }
                else if (choice == "2")
                {
                    Console.Write("Enter name or skill to search: ");
                    var query = Console.ReadLine();
                    SearchCandidates(query);
                }
                else if (choice == "3")
                {
                    Console.WriteLine("Exiting the application.");
                    break;
                }
                else
                {
                    Console.WriteLine("Invalid choice. Please try again.");
                }
            }
        }

        static Candidate CollectCandidateInfo()
        {
            Console.Write("Enter your name: ");
            var name = Console.ReadLine();

            string email;
            do
            {
                Console.Write("Enter your email: ");
                email = Console.ReadLine();
            } while (!ValidateEmail(email));

            string phone;
            do
            {
                Console.Write("Enter your phone number (e.g., +1234567890): ");
                phone = Console.ReadLine();
            } while (!ValidatePhone(phone));

            Console.Write("Do you have any certifications? (yes/no): ");
            var hasCertification = Console.ReadLine().ToLower() == "yes";
            List<string> certifications = new List<string>();

            if (hasCertification)
            {
                Console.Write("Enter your certifications (comma-separated): ");
                certifications = Console.ReadLine().Split(',').Select(c => c.Trim()).ToList();
            }

            int yearsOfExperience;
            do
            {
                Console.Write("Enter your years of experience: ");
            } while (!int.TryParse(Console.ReadLine(), out yearsOfExperience) || yearsOfExperience < 0);

            Console.Write("Enter your skills (comma-separated): ");
            var skills = Console.ReadLine().Split(',').Select(s => s.Trim()).ToList();

            return new Candidate
            {
                Name = name,
                Email = email,
                Phone = phone,
                HasCertification = hasCertification,
                Certifications = certifications,
                YearsOfExperience = yearsOfExperience,
                Skills = skills
            };
        }

        static void SaveCandidateInfo(Candidate candidate)
        {
            List<Candidate> candidates = new List<Candidate>();

            if (File.Exists(DATA_FILE))
            {
                var jsonData = File.ReadAllText(DATA_FILE);
                candidates = JsonConvert.DeserializeObject<List<Candidate>>(jsonData);
            }

            candidates.Add(candidate);
            var newJsonData = JsonConvert.SerializeObject(candidates, Formatting.Indented);
            File.WriteAllText(DATA_FILE, newJsonData);
        }

        static void SearchCandidates(string query)
        {
            if (!File.Exists(DATA_FILE))
            {
                Console.WriteLine("No candidates found.");
                return;
            }

            var jsonData = File.ReadAllText(DATA_FILE);
            var candidates = JsonConvert.DeserializeObject<List<Candidate>>(jsonData);

            var results = candidates.Where(c => c.Name.IndexOf(query, StringComparison.OrdinalIgnoreCase) >= 0 ||
                                                 c.Skills.Any(skill => skill.IndexOf(query, StringComparison.OrdinalIgnoreCase) >= 0)).ToList();

            if (results.Count > 0)
            {
                Console.WriteLine("\nSearch Results:");
                foreach (var candidate in results)
                {
                    Console.WriteLine($"Name: {candidate.Name}, Email: {candidate.Email}, Phone: {candidate.Phone}, " +
                                      $"Experience: {candidate.YearsOfExperience} years, Skills: {string.Join(", ", candidate.Skills)}, " +
                                      $"Has Certifications: {(candidate.HasCertification ? "Yes" : "No")}, " +
                                      $"Certifications: {(candidate.Certifications.Count > 0 ? string.Join(", ", candidate.Certifications) : "None")}");
                }
            }
            else
            {
                Console.WriteLine("No candidates found matching your query.");
            }
        }

        static bool ValidateEmail(string email)
        {
            var emailRegex = new System.Text.RegularExpressions.Regex(@"^[^@]+@[^@]+\.[^@]+$");
            return emailRegex.IsMatch(email);
        }

        static bool ValidatePhone(string phone)
        {
            var phoneRegex = new System.Text.RegularExpressions.Regex(@"^\+?[1-9]\d{1,14}$");
            return phoneRegex.IsMatch(phone);
        }

        class Candidate
        {
            public string Name { get; set; }
            public string Email { get; set; }
            public string Phone { get; set; }
            public bool HasCertification { get; set; }
            public List<string> Certifications { get; set; }
            public int YearsOfExperience { get; set; }
            public List<string> Skills { get; set; }
        }
    }
}

File: javascript/hiring_survey.js
const fs = require('fs');
const readline = require('readline');

const DATA_FILE = 'candidates.json';

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function validateEmail(email) {
    const emailRegex = /^[^@]+@[^@]+\.[^@]+$/;
    return emailRegex.test(email);
}

function validatePhone(phone) {
    const phoneRegex = /^\+?[1-9]\d{1,14}$/;
    return phoneRegex.test(phone);
}

function validateYearsOfExperience(years) {
    return /^\d+$/.test(years) && parseInt(years) >= 0;
}

function collectCandidateInfo() {
    return new Promise((resolve) => {
        rl.question("Enter your name: ", (name) => {
            rl.question("Enter your email: ", (email) => {
                while (!validateEmail(email)) {
                    console.log("Invalid email format. Please try again.");
                    email = rl.question("Enter your email: ");
                }

                rl.question("Enter your phone number (e.g., +1234567890): ", (phone) => {
                    while (!validatePhone(phone)) {
                        console.log("Invalid phone number format. Please try again. Use the format: +1234567890");
                        phone = rl.question("Enter your phone number (e.g., +1234567890): ");
                    }

                    rl.question("Do you have any certifications? (yes/no): ", (hasCertification) => {
                        let certifications = [];
                        if (hasCertification.toLowerCase() === "yes") {
                            rl.question("Enter your certifications (comma-separated): ", (certs) => {
                                certifications = certs.split(',').map(cert => cert.trim());
                                askExperienceAndSkills(name, email, phone, certifications, resolve);
                            });
                        } else {
                            askExperienceAndSkills(name, email, phone, certifications, resolve);
                        }
                    });
                });
            });
        });
    });
}

function askExperienceAndSkills(name, email, phone, certifications, resolve) {
    rl.question("Enter your years of experience: ", (yearsOfExperience) => {
        while (!validateYearsOfExperience(yearsOfExperience)) {
            console.log("Invalid input. Please enter a valid number of years.");
            yearsOfExperience = rl.question("Enter your years of experience: ");
        }

        rl.question("Enter your skills (comma-separated): ", (skills) => {
            const candidateInfo = {
                name,
                email,
                phone,
                hasCertification: certifications.length > 0,
                certifications,
                yearsOfExperience: parseInt(yearsOfExperience),
                skills: skills.split(',').map(skill => skill.trim())
            };

            console.log("\n=== Candidate Information ===");
            console.log(`Name: ${candidateInfo.name}`);
            console.log(`Email: ${candidateInfo.email}`);
            console.log(`Phone: ${candidateInfo.phone}`);
            console.log(`Has Certifications: ${candidateInfo.hasCertification ? 'Yes' : 'No'}`);
            console.log(`Certifications: ${certifications.length > 0 ? certifications.join(', ') : 'None'}`);
            console.log(`Years of Experience: ${candidateInfo.yearsOfExperience}`);
            console.log(`Skills: ${candidateInfo.skills.join(', ')}`);

            resolve(candidateInfo);
        });
    });
}

function saveCandidateInfo(candidate) {
    let data = [];
    if (fs.existsSync(DATA_FILE)) {
        const fileData = fs.readFileSync(DATA_FILE);
        data = JSON.parse(fileData);
    }

    data.push(candidate);
    fs.writeFileSync(DATA_FILE, JSON.stringify(data, null, 4));
}

function searchCandidates(query) {
    if (!fs.existsSync(DATA_FILE)) {
        console.log("No candidates found.");
        return;
    }

    const fileData = fs.readFileSync(DATA_FILE);
    const candidates = JSON.parse(fileData);

    const results = candidates.filter(c =>
        c.name.toLowerCase().includes(query.toLowerCase()) ||
        c.skills.some(skill => skill.toLowerCase().includes(query.toLowerCase()))
    );

    if (results.length > 0) {
        console.log("\nSearch Results:");
        results.forEach(candidate => {
            console.log(`Name: ${candidate.name}, Email: ${candidate.email}, Phone: ${candidate.phone}, ` +
                `Experience: ${candidate.yearsOfExperience} years, Skills: ${candidate.skills.join(', ')}, ` +
                `Has Certifications: ${candidate.hasCertification ? 'Yes' : 'No'}, ` +
                `Certifications: ${candidate.certifications.length > 0 ? candidate.certifications.join(', ') : 'None'}`);
        });
    } else {
        console.log("No candidates found matching your query.");
    }
}

function main() {
    const menu = () => {
        console.log("\nMenu:");
        console.log("1. Fill out the survey");
        console.log("2. Search responses");
        console.log("3. Exit");

        rl.question("Choose an option: ", (choice) => {
            if (choice === '1') {
                collectCandidateInfo().then(candidateInfo => {
                    saveCandidateInfo(candidateInfo);
                    console.log("Candidate information saved successfully.");
                    menu(); // Show menu again after saving
                });
            } else if (choice === '2') {
                rl.question("Enter name or skill to search: ", (query) => {
                    searchCandidates(query);
                    menu(); // Show menu again after searching
                });
            } else if (choice === '3') {
                console.log("Exiting the application.");
                rl.close();
            } else {
                console.log("Invalid choice. Please try again.");
                menu(); // Show menu again for invalid choice
            }
        });
    };

    menu(); // Start the menu
}

main();